libs/langchain/langchain/chains/pebblo_retrieval/base.py

```diff
@@ -0,0 +1,315 @@
+"""
+Pebblo Retrieval Chain with Identity & Semantic Enforcement for question-answering
+against a vector database.
+"""
+
+import logging
+from typing import Any, List, Optional, Union
+
+from langchain_community.vectorstores import Pinecone, Qdrant
+from langchain_core.callbacks import (
+    AsyncCallbackManagerForChainRun,
+    CallbackManagerForChainRun,
+)
+from langchain_core.documents import Document
+from langchain_core.language_models import BaseLanguageModel
+from langchain_core.pydantic_v1 import Field, validator
+from langchain_core.vectorstores import VectorStoreRetriever
+
+from langchain.chains.pebblo_retrieval.models import AuthContext, SemanticContext
+from langchain.chains.question_answering import load_qa_chain
+from langchain.chains.retrieval_qa.base import RetrievalQA
+
+logger = logging.getLogger(__name__)
+
+SUPPORTED_VECTORSTORES = [Pinecone, Qdrant]
+
+AUTH_IDENTITY_OVERRIDE_MSG = (
+    "'authorized_identities' filter in 'search_kwargs' will be overridden when using "
+    "PebbloRetrievalQA chain with 'auth_context'. Please avoid adding it in "
+    "'search_kwargs'."
+)
+
+SEMANTIC_TOPIC_OVERRIDE_MSG = (
+    "'pebblo_semantic_topics' filter in 'search_kwargs' will be overridden when using "
+    "PebbloRetrievalQA chain with 'semantic_context'. Please avoid adding it in "
+    "'search_kwargs'."
+)
+
+SEMANTIC_ENTITY_OVERRIDE_MSG = (
+    "'pebblo_semantic_entities' filter in 'search_kwargs' will be overridden when "
+    "using PebbloRetrievalQA chain with 'semantic_context'. Please avoid adding it in "
+    "'search_kwargs'."
+)
+
+
+class PebbloRetrievalQA(RetrievalQA):
+    """
+    Retrieval Chain with Identity & Semantic Enforcement for question-answering
+    against a vector database.
+    """
+
+    retriever: VectorStoreRetriever = Field(exclude=True)
+    """VectorStore to use for retrieval."""
+
+    auth_context: Optional[AuthContext] = Field(None, exclude=True)
+    """Authentication context for identity enforcement."""
+
+    semantic_context: Optional[SemanticContext] = Field(None, exclude=True)
```

Comment:
@srics What do you suggest here: a Pydantic model called SemanticContext, or two separate lists (topics_to_deny and entities_to_deny)?

This is w.r.t.: https://github.com/daxa-ai/pebblo/pull/307#discussion_r1553237819